cep-recipe-jinja.md

```diff
@@ -0,0 +1,527 @@
+is used in the new recipe format, what Jinja functions are available and how
+variables can be set and used.
+
+## Jinja in the new recipe format
+
+The new recipe format uses a subset of Jinja. Specifically, only "variable"
+expressions are allowed (no blocks such as `set`, `for` loops, `if`/`else`
+blocks, ...).
+
+A Jinja expression in the new recipe format looks like the following:
+
+`${{ version }}`
+
+Or if a function is involved:
+
+`${{ compiler('c') }}`
+
+Jinja expressions are also used in `if` statements and in the `skip` field of a
+recipe. In both instances, the `${{ ... }}` syntax is omitted.
+
+E.g.:
+
+```yaml
+build:
+  skip:
+    - osx  # This is a Jinja expression!
+
+requirements:
+  build:
+    - if: win and cuda  # This is a Jinja expression!
+      then:
+        - cudatoolkit
+```
+
+## Variables in the recipe
+
+The variables that are available in the Jinja context in the recipe come from
+two sources: the "variant configuration" file or the "context" section of the
+recipe.
+
+### The context section
+
+The context is a dictionary at the top-level of a recipe that maps keys to
+scalar values. The keys can be used by accessing them as variables in the Jinja
+expressions. For example:
+
+```yaml
+context:
+  version: "1.0.5"
+
+package:
+  version: ${{ version }}
+```
+
+Context evaluation must happen from top-to-bottom. That means a later value can
+reference an earlier one like so:
+
+```yaml
+context:
+  version: "1.0.5"
+  name_and_version: "pkg_${{ version | replace('.', '_') }}"  # evaluates to "pkg_1_0_5"
+```
+
+### Variables from variant configuration
+
+Any variable specified in the variant configuration file can be used in a recipe
+by using it in a Jinja expression. For example, with a variant config like:
+
+```yaml
+cuda:
+  - "have_cuda"
+  - "no_cuda"
+```
+
+This value can be used as follows, for example in an inline `if` expression:
+
+```yaml
+requirements:
+  host:
+    - ${{ "cudatoolkit" if cuda == "have_cuda" }}
+```
+
+### Default variables in the recipe
+
+Several variables are globally available in the recipe, based on the `target_platform` and `build_platform`:
+
+- `target_platform` is a string that represents the platform for which the package is built. It is a string of the form `os-arch` (e.g. `linux-64`, `osx-64`, `win-64`, `linux-aarch64`, ...).
+- `build_platform` is a string that represents the platform on which the package is built. Same format as `target_platform`.
+- `linux`, `osx`, `win`, `emscripten`: These are boolean variables that are `true` if the target platform is a Linux, macOS, Unix, or Windows platform, respectively. Note that this is the first part of the `target_platform` string.
+- `x86_64`, `aarch64`, `armv7l`, `ppc64le`, `s390x`, `sparc64`, `riscv64`: These are boolean variables that are `true` if the target platform is the respective architecture. Note that, except for `x86_64`, these are the second part of the `target_platform` string.
```

Comment:
This does not capture the `osx-arm64`/`win-arm64` cases.
Not a huge deal, but it would be good to mention here how those are handled.

Reply:
added `arm64`.